File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/message/digitaltwin/ReadPropertyHistoryValuesRequest.java
package org.opcfoundation.webserver.service.message.digitaltwin;

import org.eclipse.milo.opcua.stack.core.types.UaStructuredType;
import org.eclipse.milo.opcua.stack.core.types.structured.ReadAtTimeDetails;
import org.eclipse.milo.opcua.stack.core.types.structured.ReadProcessedDetails;
import org.eclipse.milo.opcua.stack.core.types.structured.ReadRawModifiedDetails;
import org.opcfoundation.webserver.types.digitaltwin.ObjectServiceContext;
import org.springframework.lang.Nullable;

public class ReadPropertyHistoryValuesRequest {
    private final String id;
    private final String propertyToRead;
    private final UaStructuredType details;
    private final ObjectServiceContext context;

    public ReadPropertyHistoryValuesRequest(
            ObjectServiceContext context,
            String propertyToRead,
            UaStructuredType details)
    {
        this.id = context.getObjectId().getId();
        this.propertyToRead = propertyToRead;
        this.details = details;
        this.context = context;
    }

    public String getId() {
        return id;
    }

    public String getPropertyToRead() {
        return propertyToRead;
    }

    public @Nullable ReadRawModifiedDetails getReadRawDetails() {
        if (details instanceof ReadRawModifiedDetails) return (ReadRawModifiedDetails)details;
        return null;
    }

    public @Nullable ReadAtTimeDetails getReadAtTimeDetails() {
        if (details instanceof ReadAtTimeDetails) return (ReadAtTimeDetails)details;
        return null;
    }

    public @Nullable ReadProcessedDetails getReadProcessedDetails() {
        if (details instanceof ReadProcessedDetails) return (ReadProcessedDetails)details;
        return null;
    }

    public ObjectServiceContext getContext() {
        return context;
    }
}

File: opcua-webservice-node/source/src/server/service/message/digitialtwin/ReadPropertyHistoryValuesRequest.ts
import { DataTypeIds, UaExtensionObject, UaNodeId, UaReadAtTimeDetails, UaReadRawModifiedDetails,UaReadProcessedDetails } from "opcua-webapi-ts";
import { ObjectServiceContext } from "../../../types/digitaltwin/ObjectServiceContext";

export class ReadPropertyHistoryValuesRequest {
    private readonly _id: string;
    private readonly _propertyToRead: string;
    private readonly _details: UaExtensionObject;
    private readonly _context: ObjectServiceContext;

    constructor(
        context: ObjectServiceContext,
        propertyToRead: string,
        details: UaExtensionObject) {
        this._id = context.objectId.id;
        this._propertyToRead = propertyToRead;
        this._details = details;
        this._context = context;
    }

    get id(): string {
        return this._id;
    }

    get propertyToRead(): string {
        return this._propertyToRead;
    }

    get details(): UaExtensionObject {
        return this._details;
    }

    get readRawDetails(): UaReadRawModifiedDetails | null {
        if (!this._details.typeId.equal(UaNodeId.from(DataTypeIds.ReadRawModifiedDetails))) {
            return null;
        }

        return UaReadRawModifiedDetails.fromExtensionObject(this._details);
    }

    get readAtTimeDetails(): UaReadAtTimeDetails | null {
        if (!this._details.typeId.equal(UaNodeId.from(DataTypeIds.ReadAtTimeDetails))) {
            return null;
        }

        return UaReadAtTimeDetails.fromExtensionObject(this._details);
    }

    get readProcessedDetails(): UaReadProcessedDetails | null {
        if (!this._details.typeId.equal(UaNodeId.from(DataTypeIds.ReadProcessedDetails))) {
            return null;
        }

        return UaReadProcessedDetails.fromExtensionObject(this._details);
    }

    get context(): ObjectServiceContext {
        return this._context;
    }
}